src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php

```diff
@@ -126,18 +126,17 @@ private function attemptSwitchUser(Request $request, $username)
             throw new \LogicException(sprintf('You are already switched to "%s" user.', $token->getUsername()));
         }
 
-        if (false === $this->accessDecisionManager->decide($token, array($this->role))) {
-            $exception = new AccessDeniedException();
-            $exception->setAttributes($this->role);
+        $username = $request->get($this->usernameParameter);
+        $user = $this->provider->loadUserByUsername($username);
 
-            throw $exception;
+        if (false === $this->accessDecisionManager->decide($token, array($this->role), $user)) {
+            throw new AccessDeniedException();
         }
 
         if (null !== $this->logger) {
             $this->logger->info('Attempting to switch to user.', array('username' => $username));
```

Comment:
Won't this give the wrong username as result now?

Reply:
I wonder why we would need to do the change on line 129 instead.

Reply:
I don't think it will give the wrong username, but I did notice that I removed the line that sets attributes on the exception, so I'll add that back in.

Reply:
@iltar, my apologies, the code I submitted did indeed change the username for JSON requests.  The problem stemmed from me originally starting from the 2.7 branch and then rebasing to master, and I ended up changing some code from 2.7 that was different in master without realizing... in any case, lesson learned and thanks for pointing this out